assembl/views/api2/discussion.py

```diff
@@ -667,6 +667,143 @@ def multi_module_csv_export(request):
     return csv_response_multiple_sheets(results, fieldnames, as_buffer=as_buffer)
 
 
+@view_config(context=InstanceContext, name="users-export",
+             ctx_instance_class=Discussion, request_method='GET',
+             permission=P_DISC_STATS)
+def users_csv_export(request):
+    import assembl.models as m
+    from collections import defaultdict
+
+    ID = "id"
+    NAME = "Nom prénom"
+    EMAIL = "Mail"
+    USERNAME = "Pseudo"
+    CREATION_DATE = "Date de création du compte"
+    FIRST_VISIT = "Date de la première connexion"
+    LAST_VISIT = "Date de la dernière connexion"
+    TOP_POSTS = "Nombre de Top posts"
+    TOP_POST_REPLIES = "Nombre de réponses reçues aux Top posts"
+    POSTS = "Nombre de posts"
+    POSTS_REPLIES = "Nombre de réponses reçues aux posts"
+    TOTAL_POSTS = "Nombre total de posts"
+    AGREE_GIVEN = '''Nombre de mentions "D'accord" donné'''
+    DISAGREE_GIVEN = '''Nombre de mentions "Pas d'accord" donné'''
+    DONT_UNDERSTAND_GIVEN = '''Nombre de mentions "Pas tout compris" donné'''
+    MORE_INFO_GIVEN = '''Nombre de mentions "SVP + d'infos" donné'''
+    AGREE_RECEIVED = '''Nombre de mentions "D'accord" reçu'''
+    DISAGREE_RECEIVED = '''Nombre de mentions "Pas d'accord" reçu'''
+    DONT_UNDERSTAND_RECEIVED = '''Nombre de mentions "Pas tout compris" reçu'''
+    MORE_INFO_RECEIVED = '''Nombre de mentions "SVP + d'infos" reçu'''
+    THEMATICS = "Liste des thématiques dans lesquelles il a contribué"
+    fieldnames = [
+        NAME, EMAIL, USERNAME, CREATION_DATE, FIRST_VISIT, LAST_VISIT, TOP_POSTS, TOP_POST_REPLIES, POSTS,
+        POSTS_REPLIES, TOTAL_POSTS, AGREE_GIVEN, DISAGREE_GIVEN, DONT_UNDERSTAND_GIVEN, MORE_INFO_GIVEN, AGREE_RECEIVED,
+        DISAGREE_RECEIVED, DONT_UNDERSTAND_RECEIVED, MORE_INFO_RECEIVED, THEMATICS
+    ]
+    LIKE_SENTIMENT = 'sentiment:like'
+    DISLIKE_SENTIMENT = 'sentiment:disagree'
+    DONT_UNDERSTAND_SENTIMENT = 'sentiment:dont_understand'
+    MORE_INFO_SENTIMENT = 'sentiment:more_info'
+
+    discussion = request.context._instance
+    db = discussion.db
+    user_prefs = LanguagePreferenceCollection.getCurrent()
+
+    start, end, interval = get_time_series_timing(request)
+    has_anon = asbool(request.GET.get('anon', False))
+
+    posts = db.query(m.Post
+        ).filter(m.Post.discussion_id == discussion.id
+        ).filter(m.Post.creation_date > start
+        ).filter(m.Post.creation_date < end
+        ).order_by(m.Post.id).all()
+
+    sentiment_counts = db.query(m.Post.id, m.SentimentOfPost.actor_id, m.SentimentOfPost.type).join(m.SentimentOfPost).filter(
+        m.Post.id.in_([post.id for post in posts]),
+        m.SentimentOfPost.tombstone_condition()
+    ).group_by(m.Post.id, m.SentimentOfPost.actor_id, m.SentimentOfPost.type).all()
+
+    # Sort sentiments for better access depending of post/user
+    sentiments_given_by_user = defaultdict(lambda: defaultdict(int))
```

Comment:
So, these things can be calculated by using the `count` func and the `group_by` in the query above.

Reply:
@ayyazdaniaryan If I was only looking for the count of sentiment per type per post, yes it would be easier. But I also have to determine the count of sentiment per type per user so it would require another query to get this info. Instead, I fetch all sentiments (only access DB once) then I process the data to get both per post and per user.